### ISSUE TITLE
Introduce an option to disable changing white space within source lines

### ISSUE THREAD

**Issue by stevenyvr987** [mvdan/sh#168]
I'm starting to use shfmt (via Visual Studio Code editor), but one behaviour that prevents me from using it wholeheartedly is the way shfmt changes the whitespace inside source lines. I wonder if there can be an option to disable this behaviour, but it would continue to change the white spaces at the beginning or end of the line.

The motivation is to not mess up custom alignment of code, creating layout that might be used to help the code read better.

**Comment by mvdan** [comment]
Could you give a few examples, please?

Note that code is aligned sometimes already, such as with comments spanning multiple lines.

Adding a mode that leaves non-indenting whitespace alone may seem interesting, but in my opinion it would cripple most of the use of the tool, rendering it to little more than an indentation enforcer - which many editors already do.

I'm trying to think to what degree the style you have could be adopted into the tool.

**Comment by stevenyvr987** [comment]
I can provide a couple of examples. The first is a cluster of one-line function definitions, with details abstracted out to show just layout issues.

Before:

```
aaa()  { test ttt  && fff     "$1"    "$2" | ggg  "$3" "$4"     ; } # aaa
aaaa() { test tttt && ffff "xxx$1" "yyy$2" | gggg      "$4" "$5"; } # aaaa

```
* Function names differ in length, but I align the beginning of the function body (left curly brace).

* The functions each begin with a test guard; they differ in details and so are of differing lengths, but I align the short-circuit operators (&&) so that the test guards are more easily recognized.

* The main action in each function is a pipeline of two function calls; each function call differ in length, but I align the pipe operators to better show their first and second components.

* Also, each function call differ in how they consume the function arguments, but I align the positions of the arguments to show commonalities and differences.

After processing with 'shfmt -i 2 -p -bn -ci':

```
aaa() { test ttt && fff "$1" "$2" | ggg "$3" "$4"; }
aaaa() { test tttt && ffff "xxx$1" "yyy$2" | gggg "$4" "$5"; }
```

* After shfmt'ing, the layout is more canonically correct, but I feel a certain amount of understanding gained through the former layout is gone. The differences would be more stark if there were more lines involved and if the details in each line were more complicated.

* Also, the small comments at the end of each line are deleted, which is alarming; this seems to be a different issue, maybe not POSIX-compliant?

The second example is more specific, involving the case command.

Before:

```
case $1 in
(some|day)            aaa  "$1" "2"     ;;
(we|will|be|together) aaaa "$1"     "3" ;;
esac
```

After processing with 'shfmt -i 2 -p -bn -ci':

```
case $1 in
some | day) aaa "$1" "$2" ;;
we | will | be | together) aaaa "$1" "3" ;;
esac
```

* The patterns differ in length, but I wanted to align the beginning of each one-line function call.

* Similar to the first example, the function calls consume arguments differently, but I wanted to align them to show commonalities.

* Also, I wanted balanced parenthesis in the patterns, but shfmt enforces the unbalanced option. I think shfmt should respect the original choice, balanced or unbalanced. (If the original choice is mixed, then leave it be, or enforce the choice of the first pattern.)

I should confirm there is value in shfmt in not making any format changes, because it would still issue warnings from the parser stage. For example, in my first-time trial of shfmt on a 200-line shell script I am writing, it was able to issue a legitimate warning that got missed by shellcheck, which I use aggressively.

Perhaps, the -s simplify option could be deepened. Currently, shfmt seems to simplify white space patterns within lines by default; if -s is used, shfmt will go further to simplify certain syntactical patterns. Perhaps a second simply option could be introduced to simplify or not white space patterns. In order for shfmt to do its job, it should reformat source code to reach syntax correctness, and then simplify white space if enabled and simplify syntax patterns if enabled. It would be good if either simplication could be done exclusively, but if not then so be it.

**Comment by mvdan** [comment]
> For example, in my first-time trial of shfmt on a 200-line shell script I am writing, it was able to issue a legitimate warning that got missed by shellcheck, which I use aggressively.

`shfmt` has no notion of warnings, it is not a linter. What did it catch? It should error on invalid programs, and valid programs that aren't parseable statically such as `let "$cannot_know_expression"`.

> and then simplify white space if enabled and simplify syntax patterns if enabled.

Whitespace changes have nothing to do with simplification. Simplification is done at the AST level, such as removing obviously redundant parentheses or subshells.

What happens with whitespace (both the indenting kind and the other kind) is different. Spaces are tabs are ignored (they're just recorded in the position information). Only newlines matter to the code that formats code.

The whitespace placed that isn't indentation simply follows rules. For example, all places that need a whitespace will have one (e.g. in between words). Whether the original program had a tab or multiple spaces in there does not matter.

After you have described your two examples above, I'm convinced that there is no way that `shfmt` could get this right. This is simply too subjective and impossible for a program following simple rules to do.

You proposed a mode that leaves whitespace that isn't indentation alone. Let me propose a narrower feature, which would also be simpler to implement. A mode that, when writing a space, would write N spaces if the original program had N spaces.

That should work nicely with your examples above, except with the case of spaces right before a semicolon. However, you could re-write your program as:

```
aaa()  { test ttt  && fff     "$1"    "$2" | ggg  "$3" "$4";      } # aaa
aaaa() { test tttt && ffff "xxx$1" "yyy$2" | gggg      "$4" "$5"; } # aaaa
```

Keeping *all* whitespace, no matter the context, would be nearly impossible. For example, would that also mean keeping newlines where they are? What about spaces that should turn into newlines? What about tabs? It opens a can of worms that I simply don't think is worth it for a program that has to format source code after all.

**Comment by stevenyvr987** [comment]
> `shfmt` has no notion of warnings, it is not a linter. What did it catch? It should error on invalid programs, and valid programs that aren't parseable statically such as `let "$cannot_know_expression"`.

I saw a warning box drop down in the editor, read the warning, changed the code accordingly and then moved on. This was within seconds of using `shfmt` for the first time. I can't remember the warning and made no notes on it, which I now regret. The warning box must be an artifact of the VS Code plugin, in how it represented some information from `shfmt`.

> Let me propose a narrower feature, which would also be simpler to implement. A mode that, when writing a space, would write N spaces if the original program had N spaces.

Yes, I think that would do it for the most part. A white space token would be represented canonically in the AST, and would be represented in the output canonically by default, or in the 'original' format if an option is enabled. If there are some edge cases for `shfmt` changing the final layout, then I the user can accept the result, or I can rewrite the source code.

Thanks for the consideration!

**Comment by mvdan** [comment]
Will work on this at some point soon. No promises that this will definitely be added - perhaps it turns out to be too complex or not practical.

> A white space token would be represented canonically in the AST

Just a clarification - whitespace and newlines aren't tokens in the AST, nor will they be. AST nodes do have position information though. You can infer how many whitespace were in between two nodes by seeing that they were in the same line, and by the difference in column positions between the end of one and the start of the other.

**Comment by stevenyvr987** [comment]
> No promises that this will definitely be added

Understand.

Another way is to add a directive to turn shfmt on/off. This would allow a user to turn off shfmt in sections where custom whitespace layout is really wanted (assuming the directive has been placed it at a semantic boundary of the source code).

**Comment by mvdan** [comment]
@stevenyvr987 please try https://github.com/mvdan/sh/tree/keep-padding and let me know how it goes.

Note that it may still break some paddings initially, for example if it replaces `` `foo` `` with `$(foo)`, thus adding an extra character to the column. It doesn't count the number of spaces that were written originally, it simply does its best to keep the same column numbers for certain words and tokens.

Otherwise, it seems to pass your examples okay:

```
 $ cat f && shfmt -kp f
aaa()  { test ttt  && fff     "$1"    "$2" | ggg  "$3" "$4"     ; } # aaa
aaaa() { test tttt && ffff "xxx$1" "yyy$2" | gggg      "$4" "$5"; } # aaaa

case $1 in
(some|day)            aaa  "$1" "2"     ;;
(we|will|be|together) aaaa "$1"     "3" ;;
esac
aaa()  { test ttt  && fff     "$1"    "$2" | ggg  "$3" "$4";      }
aaaa() { test tttt && ffff "xxx$1" "yyy$2" | gggg      "$4" "$5"; }

case $1 in
some | day) aaa            "$1" "2"     ;;
we | will | be | together) aaaa "$1" "3" ;;
esac
```

(the comments are gone and I have no idea why, though - seems like an unrelated issue)

**Comment by mvdan** [comment]
The comments bug has been fixed in c06f8bed68ea51a001350e2046a0717a6b0f9eaa.

**Comment by stevenyvr987** [comment]
I've downloaded the keep-padding branch and compiled it with go version 1.8.5. I've been testing it with the shell script I'm currently writing using `shfmt -i 2 -p -bn -kp`. For the most part the keep-padding option is working, but there are a few puzzling re-alignments. The following two examples are real code taken out of context, so please ignore the specifics.

Inside a case command,
```
  lsp-symbol) rq symbol     "$1" ;;
  lsp-hover)  rq hover      "$1" "$2" "$3" ;;
```
the hover case loses its first alignment, but keeps its second alignment.
```
  lsp-symbol) rq symbol    "$1" ;;
  lsp-hover) rq hover      "$1" "$2" "$3" ;;
```
In a series of one-line function definitions,
```
nbT()    {                  jn "{method:\"$1\"}" "{params:${2:-{\}}}"; }
rqT()    { _=$1 && shift && jn "$(msgidT "$_")" "$(nbT "$1" "$2")"; }
```
the ones without a reassignment of $1 and shifting lost alignment of the function call, but kept alignment with the parameters.
```
nbT()    { jn                  "{method:\"$1\"}" "{params:${2:-{\}}}"; }
rqT()    { _=$1 && shift && jn "$(msgidT "$_")" "$(nbT "$1" "$2")"; }
```
These are the two most puzzling examples I've noticed so far, and there seems to be a pattern but I can't quite vocalized it. I'll keep using the new branch and see whether there are others.

Also, the usage message is missing a help note for the new -kp option.

**Comment by mvdan** [comment]
Ah yes, words (such as the arguments in calls or simple commands) aren't being aligned properly. Thanks for spotting.

**Comment by mvdan** [comment]
@stevenyvr987 the issue should now be fixed - please try again.

**Comment by stevenyvr987** [comment]
I've recloned the branch and rebuilt it. It seems to be working for all the areas in my shell script that depend on custom padding!

However, there is something funny going on with inline comments with function definitions.
```
f() { xxxxxxxxxxxxxx; } # f
g() { :; } # g
h() { # h
  :
}
```
becomes
```
f() { xxxxxxxxxxxxxx; } # f
g() { :; }              # g
h() {              # h
  :
}
```
The amount of padding between the end of the function definition and beginning of a comment seems to depend on the previous line.

**Comment by mvdan** [comment]
That's sort of working as intended, as consecutive inline comments are aligned.

However, that third line is wrong - need to fix that.

**Comment by mvdan** [comment]
Fixed in 6230004519d07a7199b700887c15e5ff380be75d. Thanks for spotting so many little issues and reporting them!

**Comment by mvdan** [comment]
And now I have also made it so that comment alignments (paddings too, after all) are kept too by `-kp`.

**Comment by stevenyvr987** [comment]
I've recloned, rebuilt, and testing again, using options `-i 2 -p -bn -kp`. Inline comments now seem fine under `-kp` option.

However, consider the following three one-line function definitions, with each defining a way to do a pipeline sequence using the background command &.
```
f() {   a | b    & }
g() { { a | b; } & }
h() { ( a | b )  & }
```
becomes
```
f() {   a | b &;   }
g() {          { a | b; } &; }
h() { (a  | b) &;  }
```

1. A semicolon is added to terminate the contents inside the braces, but in these three cases, the background command is a legitimate terminator and so the semicolon shouldn't be needed. Before, shellcheck showed no problems; afterwards, it complains by annotating the semicolons in red.
2. The g function got extra padding added; it looks like shfmt is trying to align the pipe command with the end brace, which is a point of alignment.
3. The h function got the padding adjacent to the parens omitted, which I'm guessing shfmt is enforcing a higher-priority rule, but shfmt did alignment the pipe command, which is cool. But I wonder if the original padding adjacent to the parens can be retained.

**Comment by mvdan** [comment]
I'll reply in parts:

I'm aware that `&` can terminate a command, but I'm surprised that shellcheck flags `&;`. I personally prefer statements to always end in either newline or `;` regardless or any other factor. Otherwise, if I show you `foo && bar & etc && xxx`, it takes some brain processing to tell what's going on (at least for me).

The extra padding in `g` seems like a bug of some sort - will investigate.

The reason that the padding in the subshell isn't being kept is because the printer never puts spaces in between parentheses. Note that when I described this new feature, I said `A mode that, when writing a space, would write N spaces if the original program had N spaces`. The reason for that is that I never want stuff like `$(     padded )` or `f   ( ) { ... }`.

However, making an exception for subshells seems okay to me. They're used similarly to blocks and we do space those (because we must), so it does imply some consistency.

**Comment by mvdan** [comment]
2 was a72e5a299c36ce49f6d49758067962893b32522b - rebased the feature branch on top.

**Comment by stevenyvr987** [comment]
I've recloned and rebuilt, and testing shows that
```
f() {   a | b    & }
g() { { a | b; } & }
h() { ( a | b )  & }
```
becomes
```
f() {   a | b &;   }
g() { { a | b; } &; }
h() { ( a | b ) &; }
```
which is almost there.

1. The exception is the misalignment of `&` in `f()` and `h()`; I'm guessing shfmt wants to enforce a padding of one between `&` and the previous token.
2. Regarding the semicolons: shellcheck is complaining about the rule https://github.com/koalaman/shellcheck/wiki/SC1045, and showing it as an error, not just a warning. The wiki description seems to indicate the terminator should be ampersand or semicolon, not both. I've always regarded `a ;` as scheduling command `a` for synchronous execution, and `a &` for asynchronous execution. `a &;` would seem to mean schedule `a` asynchronously, followed by a null command synchronously.

The following is also illuminating.
```
{ a  & b & } # brace 1
{ a && b & } # brace 2
```
becomes
```
{
  a &
       b &
}            # brace 1
{ a && b &; } # brace 2

```
3. Brace 1 maintains alignment with brace 2, except for a newline added between `a &` and `b &`; I sort of see why that happens, because shfmt layout rule is to break a horizontal sequence into a vertical one.
4. The inline comment is out by one because a semicolon is added to brace 2, but it's puzzling why one or two isn't added to brace 1.

In conclusion, I suggest semicolon not be added to ampersand; it should be one or the other, even though `&;` seems to parse. I think `&` or `;` shouldn't be regarded as a terminator, but as a scheduling operator, and in that case the combination wouldn't make sense.

**Comment by mvdan** [comment]
1. Yes, the AST has no position information for `&`. It does have position information for the last token/node in a statement (which may be `;`, `&` or anything else if there is a newline). Nothing is done to keep the padding for `;`, so I wouldn't treat `&` differently.

2. It is unfortunate that shellcheck has such a rule - there is nothing wrong with `&;`, and the language spec allows it. I disagree that `;` means synchronous execution, though. You can skip it if there is a newline, for example. `;` and `&` cannot be put in the same bag, I would say. In a similar fashion, reading `a &;` as two commands is a misunderstanding of the language.

As a more specific point, I cannot change the fact that `a & b` is converted into `a &; b`. I could do that in v3, as a breaking change, but v2 just happened and I have no intent to break backwards compatibility again any time soon :)

3. Yes, multiple statements (commands) in a single line will be split into multiple lines. `-kp` makes this a bit confusing, but otherwise it's working as intended. `-kp` is always going to make things a bit weird the first time you run `shfmt` - it is meant to keep things stable, not to do magic to fix things for you.

4. No semicolons are added in the body of the first block because there already are newlines. Semicolons are only put when the next token (e.g. a statement, `}`, etc) follows in the same line.

**Comment by mvdan** [comment]
> The alternative strategy is to ignore padding in this specfic case and have the user readjust. I'm not sure which is best.

Your point is interesting, but the line splitting and padding stuff are completely separate rules. Making them aware of each other should be possible, but it would be added complexity.

Given that this shouldn't happen often - it just happens the first time you run the tool, and also given how neither of us is sure what to do, I'll default to keeping it the way it is for now.

> Since the semicolon is now gone, I think the padding calculation should be adjusted accordingly.

This could be done if we kept paddings by counting whitespace. But we don't do that - we count columns.

If we did the opposite, then this case would break:

```
foo;  # aligned
baaar # comments

would become:

foo  # aligned
baaar # comments
```

I'd say this falls into the same category of "no good solution, and should only bother you once" as with the previous issue.

> turn off the specific hint rule [...] or use multiple lines

I would personally disable that rule in shellcheck, but it's really up to the user. Both solutions seem fine to me. If I were designing the syntax package and `shfmt` from scratch, I may have changed my mind about `&;`. But, as said before, it is unlikely that I'll do such a breaking change anytime soon.

**Comment by mvdan** [comment]
To recap - I think this feature is somewhat ready. I'm sure there will be bugs and edge cases I missed. But if it does the job OK, in its current form it's fairly easily maintainable, and I would have no problem merging into master and shipping with 2.1 :)

**Comment by stevenyvr987** [comment]
> If we did the opposite, then this case would break...

That's a good point.

> I'm sure there will be bugs and edge cases I missed.

 I'm finding it harder to find any edge cases; if there is a difference and if I don't see a bug in a few seconds of analysis, I just re-align and re-execute shfmt to find a stable point.

> I would have no problem merging into master and shipping with 2.1 :)

I would be happy with it too! I'm using shfmt quite often now, and in tandem with shellcheck; it's a good combination, SC1045 notwithstanding.

**Comment by mvdan** [comment]
FYI, I'm considering removing this feature in v4. See https://github.com/mvdan/sh/issues/658.
